kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -1126,15 +1180,35 @@ sortInjectionAndEqualsAssumesDifferentHeads
     Nothing ->
         Monad.Trans.lift (Monad.Unify.throwUnificationError UnsupportedPatterns)
     Just NotInjection -> empty
-    Just NotMatching -> return (ExpandedPattern.bottom, SimplificationProof)
+    Just NotMatching -> do
+        Monad.Trans.lift $ Monad.Unify.explainBottom
+           (Pretty.hsep
+               [ "Unification of sort injections failed due to mismatch."
+               , "This can happen either because one of them is a constructor"
+               , "or because their sort intersection is empty."
+               ]
+           )
+           first
+           second
+        return (ExpandedPattern.bottom, SimplificationProof)
     Just
         (Matching SortInjectionMatch
             { injectionHead, sort, firstChild, secondChild }
         ) -> do
             (merged, _) <- Monad.Trans.lift $
                 termMerger firstChild secondChild
             if ExpandedPattern.isBottom merged
-                then return (ExpandedPattern.bottom, SimplificationProof)
+                then do
+                    Monad.Trans.lift $ Monad.Unify.explainBottom
```

Comment:
Why does this need to be `lift`-ed?

Reply:
Because it's a `MaybeT unifier ...`, so the `lift` instance is for `MaybeT`.

Reply:
Can we define a `MonadUnify` for `MaybeT unifier`?

Reply:
I don't think we can since `MonadUnify` has the wrong kind ("thanks" to `mapVariables`).